includes/rtm/vector4f.h

```diff
@@ -2216,7 +2216,136 @@ namespace rtm
+        if (rtm_impl::static_condition<comp0 == mix4::c && comp1 == mix4::z && comp2 == mix4::d && comp3 == mix4::w>::test())
+            return vzip2q_f32(input1, input0);
+
+        // Even-numbered vector elements, consecutively
+        if (rtm_impl::static_condition<comp0 == mix4::x && comp1 == mix4::z && comp2 == mix4::a && comp3 == mix4::c>::test())
+            return vuzp1q_f32(input0, input1);
+
+        // Even-numbered vector elements, consecutively
+        if (rtm_impl::static_condition<comp0 == mix4::a && comp1 == mix4::c && comp2 == mix4::x && comp3 == mix4::z>::test())
+            return vuzp1q_f32(input1, input0);
+
+        // Odd-numbered vector elements, consecutively
+        if (rtm_impl::static_condition<comp0 == mix4::y && comp1 == mix4::w && comp2 == mix4::b && comp3 == mix4::d>::test())
+            return vuzp2q_f32(input0, input1);
+
+        // Odd-numbered vector elements, consecutively
+        if (rtm_impl::static_condition<comp0 == mix4::b && comp1 == mix4::d && comp2 == mix4::y && comp3 == mix4::w>::test())
+            return vuzp2q_f32(input1, input0);
+
+        // Even-numbered vector elements, interleaved
+        if (rtm_impl::static_condition<comp0 == mix4::x && comp1 == mix4::a && comp2 == mix4::z && comp3 == mix4::c>::test())
+            return vtrn1q_f32(input0, input1);
+
+        // Even-numbered vector elements, interleaved
+        if (rtm_impl::static_condition<comp0 == mix4::a && comp1 == mix4::x && comp2 == mix4::c && comp3 == mix4::z>::test())
+            return vtrn1q_f32(input1, input0);
+
+        // Odd-numbered vector elements, interleaved
+        if (rtm_impl::static_condition<comp0 == mix4::y && comp1 == mix4::b && comp2 == mix4::w && comp3 == mix4::d>::test())
+            return vtrn2q_f32(input0, input1);
+
+        // Odd-numbered vector elements, interleaved
+        if (rtm_impl::static_condition<comp0 == mix4::b && comp1 == mix4::y && comp2 == mix4::d && comp3 == mix4::w>::test())
+            return vtrn2q_f32(input1, input0);
+#endif // defined(RTM_NEON64_INTRINSICS)
+
+#if defined(RTM_NEON_INTRINSICS)
+        // Each component comes from the respective position of input 0 or input 1
+        if (rtm_impl::static_condition<(comp0 == mix4::x || comp0 == mix4::a) && (comp1 == mix4::y || comp1 == mix4::b) &&
+                                       (comp2 == mix4::z || comp2 == mix4::c) && (comp3 == mix4::w || comp3 == mix4::d)>::test())
+        {
+            alignas(16) constexpr uint32_t mask_u[4] = {(comp0 == mix4::x ? UINT32_MAX : 0),
+                                                        (comp1 == mix4::y ? UINT32_MAX : 0),
+                                                        (comp2 == mix4::z ? UINT32_MAX : 0),
+                                                        (comp3 == mix4::w ? UINT32_MAX : 0)};
+            const uint32x4_t mask = vld1q_u32(&mask_u[0]);
+            return vbslq_f32(mask, input0, input1);
```

Comment:
Everything here is awesome but I'm not sure about this.
While clang might be smart sometimes, msvc struggles with this: https://godbolt.org/z/Y56Y6zchG
Sadly, unlike with AVX, there's no bitwise select that takes an immediate value and there's no clean way to my knowledge to generate the required register with a single instruction (or two) :(

I think we'll have to remove this particular case as it might adversely impact performance with more naive compilers.

Reply:
Right, that's unfortunate not to have effective implementation for MSVC and therefore drop this case for every compiler in general. But here we are.